File: app/src/main/java/info/maaskant/wouttest2/detail/DetailActivity.java
package info.maaskant.wouttest2.detail;

import static java.util.Objects.requireNonNull;

import javax.inject.Inject;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;

import info.maaskant.wouttest2.R;
import info.maaskant.wouttest2.WoutTest2App;
import info.maaskant.wouttest2.model.ContentNode;
import info.maaskant.wouttest2.utils.ApplicationInstrumentation;
import io.reark.reark.utils.RxViewBinder;
import rx.android.schedulers.AndroidSchedulers;
import rx.subscriptions.CompositeSubscription;
import timber.log.Timber;

public class DetailActivity extends AppCompatActivity {

    public static final String NODE_ID_KEY = "nodeId";

    private static final String EDITOR_FRAGMENT_KEY = "editorFragment";

    private static final String VIEWER_FRAGMENT_KEY = "viewerFragment";

    @Inject
    DetailViewModel detailViewModel;

    @Inject
    ApplicationInstrumentation instrumentation;

    private String nodeId;

    private ViewerFragment viewerFragment;

    private DetailActivity.ViewBinder activityViewBinder;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        Timber.v("onCreate (hash: %s, savedInstanceState: %s)", System.identityHashCode(this),
                savedInstanceState);

        ((WoutTest2App) getApplication()).getGraph().inject(this);

        setContentView(R.layout.activity_detail);
        Toolbar toolbar = (Toolbar) findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);

        if (getIntent().hasExtra(NODE_ID_KEY)) {
            nodeId = getIntent().getStringExtra(NODE_ID_KEY);
        } else {
            Timber.e("No node identifier specified");
            finish();
            return;
        }
        Timber.v("Using node identifier " + nodeId);

        this.viewerFragment = getOrCreateViewerFragment(savedInstanceState);

        activityViewBinder = new DetailActivity.ViewBinder(this, detailViewModel);
        // TODO: Is this the right place to call this? It used to be called from ViewerFragment
        detailViewModel.subscribeToDataStore();

        detailViewModel.setContentNodeId(nodeId);

    }

    @Override
    public void onResume() {
        super.onResume();
        Timber.v("onResume");
        activityViewBinder.bind();
    }

    @Override
    public void onPause() {
        super.onPause();
        Timber.v("onPause");
        activityViewBinder.unbind();
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        Timber.v("onDestroyView");
        detailViewModel.unsubscribeFromDataStore();
        detailViewModel.dispose();
        instrumentation.getLeakTracing().traceLeakage(this);
    }

    /**
     * Retrieves a {@link ViewerFragment} from a saved instance state or creates a new instance.
     *
     * @param savedInstanceState
     *            The {@link Bundle} to load from.
     * @return A new instance if {@code savedInstanceState} is {@code null}, a restored instance
     *         from {@link #getSupportFragmentManager()} otherwise.
     */
    private ViewerFragment getOrCreateViewerFragment(Bundle savedInstanceState) {
        if (savedInstanceState == null) {
            ViewerFragment viewerFragment = new ViewerFragment();
            getSupportFragmentManager().beginTransaction().add(R.id.viewer_fragment, viewerFragment)
                    .commit();
            return viewerFragment;
        } else {
            return (ViewerFragment) getSupportFragmentManager().getFragment(savedInstanceState,
                    VIEWER_FRAGMENT_KEY);
        }
    }

    // @Override
    // public boolean onCreateOptionsMenu(Menu menu) {
    // // Inflate the menu; this adds items to the action bar if it is present.
    // getMenuInflater().inflate(R.menu.menu_main, menu);
    // return true;
    // }

    // @Override
    // public boolean onOptionsItemSelected(MenuItem item) {
    // // Handle action bar item clicks here. The action bar will
    // // automatically handle clicks on the Home/Up button, so long
    // // as you specify a parent activity in AndroidManifest.xml.
    // int id = item.getItemId();
    //
    // // noinspection SimplifiableIfStatement
    // if (id == R.id.action_settings) {
    // return true;
    // }
    //
    // return super.onOptionsItemSelected(item);
    // }

    // @Override
    // public void onBackPressed() {
    // if (!navigationFragment.onBackPressed()) {
    // super.onBackPressed();
    // }
    // }

    @Override
    public void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        getSupportFragmentManager().putFragment(outState, VIEWER_FRAGMENT_KEY, this.viewerFragment);
    }

    void setSupportActionBarTitle(String title) {
        getSupportActionBar().setTitle(title);
    }

    /**
     * Binds a {@link DetailViewModel} to a {@link DetailActivity}.
     */
    static class ViewBinder extends RxViewBinder {

        private final DetailActivity view;
        private final DetailViewModel viewModel;

        public ViewBinder(@NonNull final DetailActivity view,
                @NonNull final DetailViewModel viewModel) {
            this.view = requireNonNull(view);
            this.viewModel = requireNonNull(viewModel);
        }

        @Override
        protected void bindInternal(@NonNull final CompositeSubscription s) {
            s.add(viewModel.getContentNode().observeOn(AndroidSchedulers.mainThread())
                    .map(ContentNode::getName).subscribe(view::setSupportActionBarTitle));
        }

    }
}

File: app/src/main/java/info/maaskant/wmsnotes/android/ui/detail/DetailActivity.kt
package info.maaskant.wmsnotes.android.ui.detail

import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import androidx.activity.viewModels
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import androidx.viewpager.widget.ViewPager
import dagger.hilt.android.AndroidEntryPoint
import info.maaskant.wmsnotes.R
import info.maaskant.wmsnotes.android.app.EmergencyBrakeActivityFinisher
import info.maaskant.wmsnotes.android.client.synchronization.EmergencyBrake
import info.maaskant.wmsnotes.android.service.ApplicationServiceManager
import info.maaskant.wmsnotes.android.ui.di.DefaultFragmentFactory
import info.maaskant.wmsnotes.client.synchronization.SynchronizationTask
import info.maaskant.wmsnotes.model.CommandBus
import info.maaskant.wmsnotes.utilities.logger
import javax.inject.Inject

@AndroidEntryPoint
class DetailActivity : AppCompatActivity() {
    companion object {
        const val AGG_ID_KEY = "aggId"
        private const val EDITOR_FRAGMENT_KEY = "editorFragment"
        private const val VIEWER_FRAGMENT_KEY = "viewerFragment"
    }

    private val logger by logger()

    private val detailViewModel: DetailViewModel by viewModels()

    lateinit var detailController: DetailController

    @Inject
    lateinit var commandBus: CommandBus

    @Inject
    lateinit var emergencyBrake: EmergencyBrake

    @Inject
    lateinit var synchronizationTask: SynchronizationTask

    private lateinit var editorFragment: EditorFragment

    private lateinit var viewerFragment: ViewerFragment

    private lateinit var viewPager: ViewPager

    override fun onCreate(savedInstanceState: Bundle?) {
        logger.trace(
            "onCreate (this: {}, savedInstanceState: {})",
            System.identityHashCode(this),
            savedInstanceState
        )

        DefaultFragmentFactory.setUp(this)
        super.onCreate(savedInstanceState)

        logger.trace("DetailActivity: vm={}", detailViewModel)

        detailController = DetailController(detailViewModel, commandBus, this)

        setContentView(R.layout.detail_activity)
        setupSupportActionBar()
        setupFragmentsAndViewPager(savedInstanceState)

        this.setTitle("")

        val aggId = if (intent.hasExtra(AGG_ID_KEY)) {
            intent.getStringExtra(AGG_ID_KEY)!!
        } else {
            logger.error("No node identifier specified")
            finish()
            return
        }
        logger.trace("Using note identifier {}", aggId)

        detailViewModel.setNote(aggId)
        detailViewModel.restoreState(savedInstanceState)

        lifecycle.addObserver(detailController)
        lifecycle.addObserver(ApplicationServiceManager.ServiceBindingLifecycleObserver(this))
        lifecycle.addObserver(EmergencyBrakeActivityFinisher(emergencyBrake, this))
    }

    override fun onCreateOptionsMenu(menu: Menu?): Boolean {
        menuInflater.inflate(R.menu.detail_menu, menu)
        return true
    }

    override fun onDestroy() {
        super.onDestroy()
        logger.trace("onDestroy")
    }

    /**
     * Retrieves a [EditorFragment] from a saved instance state or creates a new instance.
     *
     * @param savedInstanceState The [Bundle] to load from.
     * @return A new instance if `savedInstanceState` is `null`, a restored instance from [.getSupportFragmentManager] otherwise.
     */
    private fun getOrCreateEditorFragment(savedInstanceState: Bundle?): EditorFragment {
        return if (savedInstanceState == null) {
            supportFragmentManager.fragmentFactory.instantiate(
                classLoader,
                EditorFragment::class.java.name
            ) as EditorFragment
        } else {
            supportFragmentManager.getFragment(
                savedInstanceState,
                EDITOR_FRAGMENT_KEY
            ) as EditorFragment
        }
    }

    /**
     * Retrieves a [ViewerFragment] from a saved instance state or creates a new instance.
     *
     * @param savedInstanceState The [Bundle] to load from.
     * @return A new instance if `savedInstanceState` is `null`, a restored instance
     * from [.getSupportFragmentManager] otherwise.
     */
    private fun getOrCreateViewerFragment(savedInstanceState: Bundle?): ViewerFragment {
        return if (savedInstanceState == null) {
            supportFragmentManager.fragmentFactory.instantiate(
                classLoader,
                ViewerFragment::class.java.name
            ) as ViewerFragment
        } else {
            supportFragmentManager.getFragment(
                savedInstanceState,
                VIEWER_FRAGMENT_KEY
            ) as ViewerFragment
        }
    }

    public override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        supportFragmentManager.putFragment(outState, EDITOR_FRAGMENT_KEY, this.editorFragment)
        supportFragmentManager.putFragment(outState, VIEWER_FRAGMENT_KEY, this.viewerFragment)
        outState.putAll(detailViewModel.getStateToSave())
    }

    override fun onOptionsItemSelected(item: MenuItem) = when (item.itemId) {
        android.R.id.home -> {
            detailController.quit()
            true
        }
        R.id.action_save -> {
            detailController.saveAndQuit()
            true
        }
        R.id.action_delete -> {
            detailController.deleteAndQuit()
            true
        }
        else -> super.onOptionsItemSelected(item)
    }

    override fun onBackPressed() {
        detailController.quit()
    }

    private fun setTitle(title: String) {
        supportActionBar!!.title = title
    }

    private fun setupFragmentsAndViewPager(savedInstanceState: Bundle?) {
        editorFragment = getOrCreateEditorFragment(savedInstanceState)
        viewerFragment = getOrCreateViewerFragment(savedInstanceState)
        viewPager = findViewById<ViewPager>(R.id.detail_view_pager).apply {
            val pagerAdapter = DetailPagerAdapter(
                supportFragmentManager,
                editorFragment,
                viewerFragment
            )
            adapter = pagerAdapter
            addOnPageChangeListener(object : ViewPager.SimpleOnPageChangeListener() {
                override fun onPageSelected(position: Int) {
                    for (i in 0 until pagerAdapter.count) {
                        val fragment = pagerAdapter.getItem(i)
                        if (fragment is OnPageSelectedListener) {
                            if (i == position) {
                                fragment.onPageSelected()
                            } else {
                                fragment.onPageDeselected()
                            }
                        }
                    }
                }
            })
        }
    }

    private fun setupSupportActionBar() {
        val toolbar = findViewById<Toolbar>(R.id.toolbar)
        setSupportActionBar(toolbar)
        supportActionBar!!.setDisplayHomeAsUpEnabled(true)
    }
}
